source/extensions/filters/http/lua/lua_filter.h

```diff
@@ -292,24 +334,79 @@ class NoopCallbacks : public Http::AsyncClient::Callbacks {
-  int requestFunctionRef() { return lua_state_.getGlobalRef(request_function_slot_); }
-  int responseFunctionRef() { return lua_state_.getGlobalRef(response_function_slot_); }
-  uint64_t runtimeBytesUsed() { return lua_state_.runtimeBytesUsed(); }
-  void runtimeGC() { return lua_state_.runtimeGC(); }
+  FilterConfig(const envoy::extensions::filters::http::lua::v3::Lua& proto_config,
+               ThreadLocal::SlotAllocator& tls, Upstream::ClusterManager& cluster_manager,
+               Api::Api& api);
+
+  PerLuaCodeSetup* perLuaCodeSetup(const std::string& name) const {
+    auto iter = per_lua_code_setups_map_.find(name);
+    if (iter != per_lua_code_setups_map_.end()) {
+      return iter->second.get();
+    } else {
+      return nullptr;
+    }
+  }
 
   Upstream::ClusterManager& cluster_manager_;
 
 private:
-  Filters::Common::Lua::ThreadLocalState lua_state_;
-  uint64_t request_function_slot_;
-  uint64_t response_function_slot_;
+  const std::map<std::string, std::string>
+  readSourceCodes(const envoy::extensions::filters::http::lua::v3::Lua& config,
+                  Api::Api& api) const;
+
+  absl::flat_hash_map<std::string, PerLuaCodeSetupPtr> per_lua_code_setups_map_;
 };
 
 using FilterConfigConstSharedPtr = std::shared_ptr<FilterConfig>;
 
+/**
+ * Route configuration for the filter.
+ */
+class FilterConfigPerRoute : public Router::RouteSpecificFilterConfig {
+public:
+  FilterConfigPerRoute(const envoy::extensions::filters::http::lua::v3::LuaPerRoute& config,
+                       ThreadLocal::SlotAllocator& tls, Api::Api& api);
+
+  bool disabled() const { return disabled_; }
+  const std::string& name() const { return name_; }
+  PerLuaCodeSetup* perLuaCodeSetup() const { return per_lua_code_setup_ptr_.get(); }
+
+private:
+  const bool disabled_;
+  const std::string name_;
+
+  PerLuaCodeSetupPtr per_lua_code_setup_ptr_{};
+};
+
+namespace {
+
+PerLuaCodeSetup* getPerLuaCodeSetup(const FilterConfig* filter_config,
+                                    const FilterConfigPerRoute* config_per_route) {
+  // 'filter_config' will never be null.
+  ASSERT(filter_config);
+  if (!filter_config) {
+    // Although this position will never be executed, we still add a check to ensure that the
+    // function is absolutely safe.
+    return nullptr;
+  }
+  if (!config_per_route) {
+    // There is no per route config and 'GLOBAL' will be used to get the default Lua code.
+    return filter_config->perLuaCodeSetup(GLOBAL_SCRIPT_NAME);
+  } else if (config_per_route->disabled()) {
+    // Filter is disabled by route config.
+    return nullptr;
+  } else if (!config_per_route->name().empty()) {
+    // Get the Lua code referenced by 'name'.
+    return filter_config->perLuaCodeSetup(config_per_route->name());
+  } else {
+    // Get the inline Lua code in the route.
+    return config_per_route->perLuaCodeSetup();
+  }
+  return nullptr;
```

Comment:
To make it clear that there is always a "catch-all" fallback, let's structure the code as below (please observe the structure, not the comments):

```cpp

// As mentioned, the most specific config wins.
if (config_per_route != nullptr) {
  if (config_per_route->disabled()) {
    return nullptr;
  }
	
  if (!config_per_route->name().empty()) {
    return filter_config->perLuaCodeSetup(config_per_route->name());
  }

  return config_per_route->perLuaCodeSetup();
}

// Catch-all.
return filter_config->perLuaCodeSetup(GLOBAL_SCRIPT_NAME);
```